### PR TITLE
Implemets category

### PR DESCRIPTION
**Pull request by Yam0r**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -149,26 +149,26 @@
...
-				</executions>
-				<configuration>
-					<configLocation>checkstyle.xml</configLocation>
-					<consoleOutput>true</consoleOutput>
-					<failsOnError>true</failsOnError>
-					<linkXRef>false</linkXRef>
-					<sourceDirectories>src</sourceDirectories>
-				</configuration>
-			</plugin>
+<!--			<plugin>-->
```
Comment (Elena-Bruyako) [inline pom.xml]: please, uncomment checkstyle first, then submit your PR again

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
please, uncomment checkstyle first, then submit your PR again

File: src/main/java/mate/academy/springintro/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+@RequestMapping("/categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @PostMapping
+    @Operation(summary = "Create a new category",
+            description = "Create a new category for books in the system")
+    public CategoryDto createCategory(@RequestBody @Valid CategoryDto categoryDto) {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/controller/CategoryController.java]: ResponseStatus.CREATED

File: src/main/java/mate/academy/springintro/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return categoryService.save(categoryDto);
+    }
+
+    @PreAuthorize("hasAuthority('USER') or hasAuthority('ADMIN')")
+    @GetMapping
+    @Operation(summary = "Get all categories",
+            description = "Retrieve a list of all available book categories")
+    public List<CategoryDto> getAll() {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/controller/CategoryController.java]: should you use pagination here?

File: src/main/java/mate/academy/springintro/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    public CategoryDto updateCategory(@PathVariable Long id,
+                                      @RequestBody @Valid CategoryDto categoryDto) {
+        return categoryService.update(id, categoryDto);
+    }
+
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @DeleteMapping("/{id}")
+    @Operation(summary = "Delete category",
+            description = "Delete an existing category by its ID")
+    public void deleteCategory(@Parameter(description = "ID of the category to delete")
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/controller/CategoryController.java:84]: ResponseStatus.NO_CONTENT

File: src/main/java/mate/academy/springintro/dto/book/BookDto.java [hunk 5]
```diff
@@ -20,4 +21,6 @@ public class BookDto {
     private String description;
 
     private String coverImage;
+
+    private List<String> categoryName;
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/dto/book/BookDto.java]: you should add `categoryIds` instead `names`. Also check your task you have to return BookDtoWithoutCategoryIds in BookController

File: src/main/java/mate/academy/springintro/dto/book/CategoryDto.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package mate.academy.springintro.dto.book;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class CategoryDto {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/dto/book/CategoryDto.java]: Let’s  use separate Dto for request and response

File: src/main/java/mate/academy/springintro/dto/book/CreateBookRequestDto.java [hunk 7]
```diff
@@ -26,4 +26,6 @@ public class CreateBookRequestDto {
     private String description;
 
     private String coverImage;
+
+    private String categoryName;
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/dto/book/CreateBookRequestDto.java]: The same. Can a book have several categories? 
Also, use NotEmpty validation

File: src/main/java/mate/academy/springintro/mapper/BookMapper.java [hunk 8]
```diff
@@ -14,4 +17,15 @@ public interface BookMapper {
...
+
+    @AfterMapping
+    default void setCategoryIds(@MappingTarget BookDto bookDto, Book book) {
+        if (book.getCategories() != null) {
+            bookDto.setCategoryName(
+                    book.getCategories().stream()
+                            .map(Category::getName)
+                            .collect(Collectors.toList())
+            );
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/mapper/BookMapper.java]:
```suggestion
        if (book.getCategories() == null) {
            return;
        }
            bookDto.setCategoryName(
                    book.getCategories().stream()
                            .map(Category::getName)
                            .toList()
            );
```

File: src/main/java/mate/academy/springintro/model/Book.java [hunk 9]
```diff
@@ -40,4 +48,15 @@ public class Book {
     private String coverImage;
     @Column(nullable = false, columnDefinition = "TINYINT")
     private boolean isDeleted;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/model/Book.java]: Don’t use the EAGER fetch type. Refactor your DB queries.

File: src/main/java/mate/academy/springintro/model/Category.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false)
+    private String name;
+
+    private String description;
+
+    @Column(nullable = false, name = "is_deleted")
+    private Boolean deleted = false;
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/model/Category.java]:
```suggestion
    private boolean deleted = false;
```
Don’t use Wrapper. It can provide `null` value.

File: src/main/java/mate/academy/springintro/model/Category.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Getter
+@Setter
+@Where(clause = "is_deleted = FALSE")
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/model/Category.java:18]: Let’s  implement soft-deleted

File: src/main/java/mate/academy/springintro/model/Category.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+@Table(name = "category")
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/model/Category.java]:
```suggestion
@Table(name = "categories")
```

File: src/main/java/mate/academy/springintro/repository/CategoryRepository.java [hunk 13]
```diff
@@ -0,0 +1,12 @@
+package mate.academy.springintro.repository;
+
+import java.util.List;
+import java.util.Optional;
+import mate.academy.springintro.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+    List<Category> findAllByDeletedFalse();
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/repository/CategoryRepository.java]:
```suggestion
    List<Category> findAll();
```

File: src/main/java/mate/academy/springintro/service/impl/BookServiceImpl.java [hunk 14]
```diff
@@ -22,19 +27,27 @@ public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
     private final BookSpecificationBuilder bookSpecificationBuilder;
+    private final CategoryRepository categoryRepository;
 
     @Override
     public BookDto save(CreateBookRequestDto requestDto) {
+        Category category = categoryRepository.findByName(requestDto.getCategoryName())
+                .orElseThrow(() -> new EntityNotFoundException("Category not found"));
+
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/service/impl/BookServiceImpl.java]: use category ID instead name

File: src/main/java/mate/academy/springintro/service/impl/BookServiceImpl.java [hunk 15]
```diff
@@ -22,19 +27,27 @@ public class BookServiceImpl implements BookService {
...
     }
 
     @Override
     public List<BookDto> findAll(Pageable pageable) {
-        return bookRepository.findAll(pageable).stream()
+        List<Book> books = bookRepository.findAll();
+        return books.stream()
                 .map(bookMapper::toBookDto)
-                .toList();
+                .collect(Collectors.toList());
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/service/impl/BookServiceImpl.java]:
```suggestion
                .toList();
```
Let’s  revert these changes

File: src/main/java/mate/academy/springintro/service/impl/BookServiceImpl.java [hunk 16]
```diff
@@ -66,4 +79,12 @@ public List<BookDto> search(BookSearchParameters params, Pageable pageable) {
                 .map(bookMapper::toBookDto)
                 .toList();
     }
+
+    @Override
+    public List<BookDto> findByCategoryId(Long categoryId) {
+        List<Book> books = bookRepository.findAllByCategoriesId(categoryId);
+        return books.stream()
+                .map(bookMapper::toBookDto)
+                .collect(Collectors.toList());
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/service/impl/BookServiceImpl.java]:
```suggestion
                .toList();
```
Fix it everywhere

File: src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+        category = categoryRepository.save(category);
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryDto update(Long id, CategoryDto categoryDto) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Category not found"));
+        category.setName(categoryDto.getName());
+        category.setDescription(categoryDto.getDescription());
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java]: use mapper for it

File: src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Category not found"));
+        category.setDeleted(true);
+        categoryRepository.save(category);
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java]: Let’s  implement soft-deleted and call method `delete` here

File: src/main/resources/db/changelog/changes/adding-category-fields-to-the-books-table.yaml [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+  - changeSet:
+      id: 3
+      author: mateacademy
+      changes:
+        - addColumn:
+            tableName: books
+            columns:
+              - column:
+                  name: categories
+                  type: varchar(255)
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/adding-category-fields-to-the-books-table.yaml]: Why do you need this column? What are you saving here?

File: src/main/resources/db/changelog/changes/creates-category-table.yaml [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+              - column:
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+                  value: 0
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/creates-category-table.yaml]:
```suggestion
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 21]
```diff
@@ -11,9 +11,15 @@ databaseChangeLog:
...
-      file: db/changelog/changes/assign-roles-to-users.yaml
\ No newline at end of file
+#  - include:
+#      file: db/changelog/changes/insert-roles.yaml
+#  - include:
+#      file: db/changelog/changes/insert-roles-of-users.yaml
+#  - include:
+#      file: db/changelog/changes/assign-roles-to-users.yaml
+#  - include:
+#      file: db/changelog/changes/creates-category-table.yaml
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: Let’s  remove commented code

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 22]
```diff
@@ -11,9 +11,15 @@ databaseChangeLog:
...
+  - include:
+      file: db/changelog/changes/insert-category.yaml
+  - include:
+      file: db/changelog/changes/adding-category-fields-to-the-books-table.yaml
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: Let’s  add `books_categories` table

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve your solution ;)

File: src/main/java/mate/academy/springintro/dto/category/CategoryRequestDto.java [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+@Getter
+@Setter
+public class CategoryRequestDto {
+    @NotBlank(message = "Name cannot be blank")
+    @Size(max = 255, message = "Name cannot exceed 255 characters")
+    private String name;
+
+    @Size(max = 500, message = "Description cannot exceed 500 characters")
+    private String description;
+
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/dto/category/CategoryRequestDto.java:17]:
```suggestion
```

File: src/main/java/mate/academy/springintro/controller/CategoryController.java [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+    public CategoryResponseDto createCategory(@RequestBody
+                                                  @Valid CategoryRequestDto categoryRequestDto) {
+        return categoryService.save(categoryRequestDto);
+    }
+
+    @PreAuthorize("hasAuthority('USER') or hasAuthority('ADMIN')")
+    @GetMapping
+    @Operation(summary = "Get all categories",
+            description = "Retrieve a list of all available book categories")
+    public List<CategoryResponseDto> getAll() {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/controller/CategoryController.java]: Let’s  use Pagination

File: src/main/java/mate/academy/springintro/mapper/BookMapper.java (hunk 8 above)
Comment (liliia-ponomarenko) [reply src/main/java/mate/academy/springintro/mapper/BookMapper.java]: not fixed

File: src/main/java/mate/academy/springintro/mapper/CategoryMapper.java [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+
+@Mapper(componentModel = "spring")
+public interface CategoryMapper {
+    CategoryResponseDto toDto(Category category);
+
+    Category toEntity(@Valid CategoryRequestDto categoryResponseDto);
+
+    void updateCategoryFromDto(@Valid CategoryRequestDto categoryResponseDto,
+                               @MappingTarget Category category);
+
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/mapper/CategoryMapper.java]:
```suggestion
```

File: src/main/java/mate/academy/springintro/model/Book.java (hunk 9 above)
Comment (liliia-ponomarenko) [reply src/main/java/mate/academy/springintro/model/Book.java]: not fixed

File: src/main/java/mate/academy/springintro/model/Category.java (hunk 11 above)
Comment (liliia-ponomarenko) [reply src/main/java/mate/academy/springintro/model/Category.java:18]: not fixed

File: src/main/java/mate/academy/springintro/repository/CategoryRepository.java [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+import mate.academy.springintro.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Modifying;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+    @Modifying
+    @Query("UPDATE Category c SET c.deleted = true WHERE c.id = :id")
+    void softDeleteById(@Param("id") Long id);
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/repository/CategoryRepository.java]: Use the same soft deleted realization that you already used for the Book.

File: src/main/java/mate/academy/springintro/service/impl/BookServiceImpl.java [hunk 27]
```diff
@@ -22,17 +25,24 @@ public class BookServiceImpl implements BookService {
...
         Book book = bookMapper.toModel(requestDto);
+        book.setCategories(new HashSet<>(categories));
         book = bookRepository.save(book);
         return bookMapper.toBookDto(book);
     }
 
     @Override
     public List<BookDto> findAll(Pageable pageable) {
-        return bookRepository.findAll(pageable).stream()
+        List<Book> books = bookRepository.findAll();
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/service/impl/BookServiceImpl.java:44]: use pagination

File: src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+@Service
+public class CategoryServiceImpl implements CategoryService {
+    private final CategoryRepository categoryRepository;
+    private final CategoryMapper categoryMapper;
+
+    @Override
+    public List<CategoryResponseDto> findAll() {
+        return categoryRepository.findAll().stream()
+                .map(categoryMapper::toDto)
+                .collect(Collectors.toList());
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java]:
```suggestion
                .toList();
```
Fix it everywhere

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 29]
```diff
@@ -12,4 +12,5 @@ databaseChangeLog:
   - include:
       file: db/changelog/changes/create-users-roles-table.yaml
   - include:
-      file: db/changelog/changes/insert-roles.yaml
+      file: db/changelog/changes/insert-category.yaml
+
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml:16]:
```suggestion
```

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix all previous comments!  
Please don’t resolve comments! Use stickers if you need to mark fixed comments

File: src/main/java/mate/academy/springintro/dto/book/BookDto.java (hunk 5 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springintro/dto/book/BookDto.java]: fix comment

File: src/main/java/mate/academy/springintro/mapper/CategoryMapper.java [hunk 30]
```diff
@@ -0,0 +1,18 @@
...
+import mate.academy.springintro.dto.category.CategoryResponseDto;
+import mate.academy.springintro.model.Category;
+import org.mapstruct.Mapper;
+import org.mapstruct.MappingTarget;
+
+@Mapper(componentModel = "spring")
+public interface CategoryMapper {
+    CategoryResponseDto toDto(Category category);
+
+    Category toEntity(@Valid CategoryRequestDto categoryResponseDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/mapper/CategoryMapper.java]: do you need `@Valid` ?

File: src/main/java/mate/academy/springintro/model/Book.java [hunk 31]
```diff
@@ -38,6 +46,18 @@ public class Book {
     private String description;
 
     private String coverImage;
-    @Column(nullable = false, columnDefinition = "TINYINT")
-    private boolean isDeleted;
+
+    @Column(nullable = false)
+    private boolean deleted = false;
+
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/Book.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/springintro/model/Book.java [hunk 32]
```diff
@@ -38,6 +46,18 @@ public class Book {
...
+    @JoinTable(
+            name = "book_category",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/Book.java]:
```suggestion
            name = "books_categories",
```

File: src/main/java/mate/academy/springintro/model/Category.java [hunk 33]
```diff
@@ -0,0 +1,32 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false)
+    private String name;
+
+    private String description;
+
+    @Column(nullable = false, name = "is_deleted")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/Category.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/mate/academy/springintro/service/impl/BookServiceImpl.java (hunk 27 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springintro/service/impl/BookServiceImpl.java:44]: fix comment

File: src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java [hunk 34]
```diff
@@ -0,0 +1,58 @@
...
+    public CategoryResponseDto getById(Long id) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Category not found"));
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryResponseDto save(@Valid CategoryRequestDto categoryResponseDto) {
+        Category category = categoryMapper.toEntity(categoryResponseDto);
+        category = categoryRepository.save(category);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java]:
```suggestion
        categoryRepository.save(category);
```

File: src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java [hunk 35]
```diff
@@ -0,0 +1,58 @@
...
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryResponseDto update(Long id, @Valid CategoryRequestDto categoryResponseDto) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Category not found"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java]: add id to message

File: src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java [hunk 36]
```diff
@@ -0,0 +1,58 @@
...
+        categoryMapper.updateCategoryFromDto(categoryResponseDto, category);
+
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        if (!categoryRepository.existsById(id)) {
+            throw new EntityNotFoundException("Category not found");
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java]: same here

File: src/main/java/mate/academy/springintro/repository/CategoryRepository.java [hunk 37]
```diff
@@ -0,0 +1,12 @@
+package mate.academy.springintro.repository;
+
+import mate.academy.springintro.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.repository.query.Param;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+    void softDeleteById(@Param("id") Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/repository/CategoryRepository.java]: do you need this method?

File: src/main/java/mate/academy/springintro/repository/CategoryRepository.java [hunk 38]
```diff
@@ -0,0 +1,12 @@
...
+
+    boolean existsByIdAndDeletedFalse(Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/repository/CategoryRepository.java]: same here

File: src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java [hunk 39]
```diff
@@ -0,0 +1,58 @@
...
+    public List<CategoryResponseDto> findAll(int page, int size) {
+        return categoryRepository.findAll().stream()
+                .map(categoryMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public CategoryResponseDto getById(Long id) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Category not found"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java]: add id to message

File: src/main/resources/db/changelog/changes/creates-category-table.yaml [hunk 40]
```diff
@@ -0,0 +1,29 @@
+databaseChangeLog:
+  - changeSet:
+      id: creates-category-table
+      author: mateacademy
+      changes:
+        - createTable:
+            tableName: category
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/creates-category-table.yaml]:
```suggestion
            tableName: categories
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/springintro/mapper/CategoryMapper.java [hunk 41]
```diff
@@ -0,0 +1,18 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.MappingTarget;
+
+@Mapper(componentModel = "spring")
+public interface CategoryMapper {
+    CategoryResponseDto toDto(Category category);
+
+    Category toEntity(CategoryRequestDto categoryResponseDto);
+
+    void updateCategoryFromDto(@Valid CategoryRequestDto categoryResponseDto,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/mapper/CategoryMapper.java]:
```suggestion
    void updateCategoryFromDto(CategoryRequestDto categoryResponseDto,
```

File: src/main/java/mate/academy/springintro/service/CategoryService.java [hunk 42]
```diff
@@ -0,0 +1,18 @@
...
+import java.util.List;
+import mate.academy.springintro.dto.category.CategoryRequestDto;
+import mate.academy.springintro.dto.category.CategoryResponseDto;
+
+public interface CategoryService {
+    List<CategoryResponseDto> findAll(int page, int size);
+
+    CategoryResponseDto getById(Long id);
+
+    CategoryResponseDto save(@Valid CategoryRequestDto categoryResponseDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/CategoryService.java]:
```suggestion
    CategoryResponseDto save(CategoryRequestDto categoryResponseDto);
```

File: src/main/java/mate/academy/springintro/service/CategoryService.java [hunk 43]
```diff
@@ -0,0 +1,18 @@
...
+
+    CategoryResponseDto update(Long id, @Valid CategoryRequestDto categoryResponseDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/CategoryService.java]:
```suggestion
    CategoryResponseDto update(Long id, CategoryRequestDto categoryResponseDto);
```

File: src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java [hunk 44]
```diff
@@ -0,0 +1,66 @@
...
+    public CategoryResponseDto getById(Long id) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Category with id "
+                        + id
+                        + " not found"));
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryResponseDto save(@Valid CategoryRequestDto categoryResponseDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java]:
```suggestion
    public CategoryResponseDto save(CategoryRequestDto categoryResponseDto) {
```

File: src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java [hunk 45]
```diff
@@ -0,0 +1,66 @@
...
+        Category category = categoryMapper.toEntity(categoryResponseDto);
+        categoryRepository.save(category);
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryResponseDto update(Long id, @Valid CategoryRequestDto categoryResponseDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java]:
```suggestion
    public CategoryResponseDto update(Long id, CategoryRequestDto categoryResponseDto) {
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/springintro/model/Category.java [hunk 46]
```diff
@@ -0,0 +1,32 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false)
+    private String name;
+
+    private String description;
+
+    @Column(nullable = false)
+    private boolean deleted = false;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/Category.java]:
```suggestion
    private boolean isDeleted = false;
```

File: src/main/java/mate/academy/springintro/model/Book.java [hunk 47]
```diff
@@ -38,6 +45,18 @@ public class Book {
     private String description;
 
     private String coverImage;
-    @Column(nullable = false, columnDefinition = "TINYINT")
-    private boolean isDeleted;
+
+    @Column(nullable = false)
+    private boolean deleted = false;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/Book.java]:
```suggestion
    private boolean isDeleted = false;
```

File: src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java [hunk 48]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    public CategoryResponseDto update(Long id, CategoryRequestDto categoryResponseDto) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Category with id "
+                        + id
+                        + " not found"));
+        categoryMapper.updateCategoryFromDto(categoryResponseDto, category);
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/CategoryServiceImpl.java]: remove empty line

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]
